src/controller/editingcontroller.js

```diff
@@ -110,6 +107,58 @@ export default class EditingController {
 		this.modelToView.on( 'selection', convertRangeSelection(), { priority: 'low' } );
 		this.modelToView.on( 'selection', convertCollapsedSelection(), { priority: 'low' } );
 
+		// Convert markers removal.
+		//
+		// Markers should be removed from the view before changes to model are applied. This is because otherwise
+		// it would be impossible to map some markers to view (if, for example, marker boundary parent got removed).
+		//
+		// `removedMarkers` keeps information which markers already has been removed to prevent removing them twice.
+		// (The second remove would not be at "the soonest moment" and it could crash.)
+		const removedMarkers = new Set();
+
+		this.listenTo( model, 'applyOperation', ( evt, args ) => {
+			// Before operation is applied...
+			const operation = args[ 0 ];
+
+			for ( const marker of model.markers ) {
+				// Check all markers, that aren't already removed...
+				if ( removedMarkers.has( marker.name ) ) {
+					continue;
+				}
+
+				const markerRange = marker.getRange();
+
+				if ( _operationAffectsMarker( operation, marker ) ) {
+					// And if the operation in any way modifies the marker, remove the marker from the view.
+					removedMarkers.add( marker.name );
+					this.modelToView.convertMarkerRemove( marker.name, markerRange );
+
+					// TODO: This stinks but this is the safest place to have this code.
+					this.model.document.differ.bufferMarkerChange( marker.name, markerRange, markerRange );
+				}
+			}
+		}, { priority: 'high' } );
+
+		// If a marker with given name was added again, remove it from `removedMarkers`. This is to prevent a bug in
+		// a situation when marker is removed, then added, then removed again (would not got removed if it is saved in `removedMarkers`).
```

Comment:
Are you sure we need this? If a marker is added, we add it on "changes done", so it should not be removed immediately, only removed from differ.

Reply:
Wow, good catch! I've added a test and it is true that it is not only unnecessary but it's even harmful.